src/libcore/str/mod.rs

```diff
@@ -1074,26 +1075,82 @@ unsafe fn cmp_slice(a: &str, b: &str, len: usize) -> i32 {
 }
 
 /*
-Section: Misc
+Section: UTF-8 validation
 */
 
+// use truncation to fit u64 into usize
+const NONASCII_MASK: usize = 0x80808080_80808080u64 as usize;
+
+/// Return `true` if any byte in the word `x` is nonascii (>= 128).
+#[inline]
+fn contains_nonascii(x: usize) -> bool {
+    (x & NONASCII_MASK) != 0
+}
+
+/// Unchecked byte slice
+unsafe fn slice_unchecked(v: &[u8], from: usize) -> &[u8] {
+    debug_assert!(from <= v.len());
+    slice::from_raw_parts(v.as_ptr().offset(from as isize),
+                          v.len() - from)
+}
+
+/// Return the first index where a nonascii byte is located,
+/// or `None` if no was found.
+fn find_nonascii(v: &[u8]) -> Option<usize> {
+    let len = v.len();
+    let ptr = v.as_ptr();
+
+    let mut offset = 0;
+    if len >= 2 * usize::BYTES {
```

Comment:
Can you extract the 2 to a const with a descriptive name about unrolling? Since I don't see any hand-unrolling here, I am guessing that the `if` statement allows the compiler to the unrolling according to the unrolling factor. This is not obvious to me. Can you also add a comment explaining?